src/widgets/index/index.ts

```diff
@@ -0,0 +1,277 @@
+  let helper: Helper | null = null;
+  let derivedHelper: DerivedHelper | null = null;
+
+  if (indexName === null) {
+    throw new Error(withUsage('The `indexName` option is required.'));
+  }
+
+  return {
+    getHelper() {
+      return helper;
+    },
+
+    getWidgets() {
+      return localWidgets;
+    },
+
+    addWidgets(widgets) {
+      if (!Array.isArray(widgets)) {
+        throw new Error(
+          withUsage('The `addWidgets` method expects an array of widgets.')
+        );
+      }
+
+      if (
+        widgets.some(
+          widget =>
+            typeof widget.init !== 'function' &&
+            typeof widget.render !== 'function'
+        )
+      ) {
+        throw new Error(
+          withUsage(
+            'The widget definition expects a `render` and/or an `init` method.'
+          )
+        );
+      }
+
+      // The routing manager widget is always added manually at the last position.
+      // By removing it from the last position and adding it back after, we ensure
+      // it keeps this position.
+      // fixes #3148
+      const lastWidget = localWidgets.pop();
+
+      localWidgets = localWidgets.concat(widgets);
+
+      if (lastWidget) {
+        // Second part of the fix for #3148
+        localWidgets = localWidgets.concat(lastWidget);
+      }
+
+      if (localInstantSearchInstance && Boolean(widgets.length)) {
+        helper!.setState(
+          localWidgets.reduce(enhanceConfiguration, {
+            ...helper!.state,
+          })
+        );
+
+        widgets.forEach(widget => {
+          if (localInstantSearchInstance && widget.init) {
+            widget.init({
+              helper: helper!,
+              parent: this,
+              instantSearchInstance: localInstantSearchInstance,
+              state: helper!.state,
+              templatesConfig: localInstantSearchInstance.templatesConfig,
+              createURL: localInstantSearchInstance._createAbsoluteURL,
+              onHistoryChange: localInstantSearchInstance._onHistoryChange,
+            });
+          }
+        });
+
+        localInstantSearchInstance.scheduleSearch();
+      }
+
+      return this;
+    },
+
+    removeWidgets(widgets) {
+      if (!Array.isArray(widgets)) {
+        throw new Error(
+          withUsage('The `removeWidgets` method expects an array of widgets.')
+        );
+      }
+
+      if (widgets.some(widget => typeof widget.dispose !== 'function')) {
+        throw new Error(
+          withUsage('The widget definition expects a `dispose` method.')
+        );
+      }
+
+      localWidgets = localWidgets.filter(
+        widget => widgets.indexOf(widget) === -1
+      );
+
+      if (localInstantSearchInstance && Boolean(widgets.length)) {
+        const nextState = widgets.reduce((state, widget) => {
+          // the `dispose` method exists at this point we already assert it
+          const next = widget.dispose!({ helper: helper!, state });
```

Comment:
`dispose` is optional on widget, so shouldn't we check if dispose exists before calling it?
Or maybe we should add spit an error message if `dispose` is not implemented on an added widget.